xds/src/main/java/io/grpc/xds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,362 @@
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.grpc.xds.internal.sds.SslContextProviderSupplier;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.util.AsciiString;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+
+/**
+ * Handles L4 filter chain match for the connection based on the xds configuration.
+ * */
+final class FilterChainMatchingHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger log = Logger.getLogger(FilterChainMatchingHandler.class.getName());
+
+  private static final AsciiString SCHEME = AsciiString.of("http");
+
+  private final GrpcHttp2ConnectionHandler grpcHandler;
+  private final FilterChainSelector selector;
+  private final ProtocolNegotiator delegate;
+
+  FilterChainMatchingHandler(
+          GrpcHttp2ConnectionHandler grpcHandler, FilterChainSelector selector,
+          ProtocolNegotiator delegate) {
+    this.grpcHandler = checkNotNull(grpcHandler, "grpcHandler");
+    this.selector = checkNotNull(selector, "selector");
+    this.delegate = checkNotNull(delegate, "delegate");
+  }
+
+  @Override
+  public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+    if (!(evt instanceof ProtocolNegotiationEvent)) {
+      super.userEventTriggered(ctx, evt);
+      return;
+    }
+    SelectedConfig config = selector.select(
+            (InetSocketAddress) ctx.channel().localAddress(),
+            (InetSocketAddress) ctx.channel().remoteAddress());
+    if (config == null) {
+      log.log(Level.FINER, "Did not find exactly one filter chain.");
+      ctx.fireExceptionCaught(
+              new IllegalStateException("Did not find exactly one filter chain."));
+      return;
+    }
+    ProtocolNegotiationEvent pne = (ProtocolNegotiationEvent)evt;
+    Attributes attr = InternalProtocolNegotiationEvent.getAttributes(pne)
+            .toBuilder().set(ATTR_SERVER_SSL_CONTEXT_PROVIDER_SUPPLIER,
+                    config.sslContextProviderSupplier).build();
+    pne = InternalProtocolNegotiationEvent.withAttributes(pne, attr);
+    ctx.pipeline().replace(this, null, delegate.newHandler(grpcHandler));
+    ctx.fireUserEventTriggered(pne);
+  }
+
+  static final class FilterChainSelector {
+    public static final FilterChainSelector NO_FILTER_CHAIN = new FilterChainSelector(
+            Collections.<FilterChain>emptyList(), null);
+
+    private final List<FilterChain> filterChainList;
+    @Nullable
+    private final SslContextProviderSupplier defaultSslContextProviderSupplier;
```

Comment:
For the default case why save `defaultSslContextProviderSupplier` instead of the defaultFilterChain? Looking at `XdsServerWrapper.DiscoveryState.updateSelector()` it could just pass `defaultFilterChain` instead of `defaultFilterChain.getSslContextProviderSupplier()` unless there is a strong reason for this inconsistency

Reply:
because `defaultFilterChain` is not consumable for default routing config, although it is ok for default `SslContextProviderSupplier`. Later on, we would add another field in the `FilterChainSelector` which is the default routing config. Default routing config is generated from parsing/watching on HCM in `defaultFilterChain`.

Reply:
> defaultFilterChain is not consumable for default routing config
...
> Default routing config is generated from parsing/watching on HCM in defaultFilterChain

I thought the routing config is always derived from the selected filterChain the same way via HCM and all that - whether the matched filterChain or the defaultFilterChain. Anyway this is not a big issue for now